src/Workspaces/Core/Portable/Shared/TestHooks/AsynchronousOperationListener.cs

```diff
@@ -94,13 +102,24 @@ private void Decrement_NoLock(AsyncToken token)
             }
         }
 
-        public Task CreateWaitTask()
+        public Task CreateWaitTask(bool willBlockOnCompletion)
         {
+            if (willBlockOnCompletion && _counter > 0)
+            {
+                _blockedOnCompletionTokenSource.Cancel();
+            }
+
             using (_gate.DisposableWait(CancellationToken.None))
             {
                 if (_counter == 0)
                 {
                     // There is nothing to wait for, so we are immediately done
+                    if (_blockedOnCompletionTokenSource.IsCancellationRequested)
```

Comment:
I am not sure I understand this. so.. this is the case where CreateWaitTask is called while _count > 0 and then later CreateWaitTask is called again once _count == 0 .. and then for this to work as epxected, it is expected someone to call CreateWaitTask again?

why, does this require to reset cancellation? once cancellation is called, shouldn't thing to stay as cancelled?

Reply:
This handles a race condition:

1. Operations exist, so the token source is marked as blocked
2. Operations complete
3. The semaphore is acquired
4. We find out that no operations need to run, so we return early

Reply:
@sharwell Is that special case avoided if we just put the cancelling of the token inside the gate? That seems easier and given it's not like the code a few lines earlier is avoiding taking the lock we're reducing any contention in practice. It also means you could just put the cancel by the code that's adding the task to _pendingTasks, which then means it's easier to audit this -- all code that adds or removes from _pendingTasks should also be manipulating this token in some way, since you're just maintaining the invariant that the token is cancelled if _pendingTasks is non-empty.

(if you _really_ want to leave it this way, you need a very long comment here.)

Reply:
It _was_ special because it needed asynchronous cancellation to avoid throwing in an attempt to re-enter the gate. I moved it inside the gate now and used `CancelAfter` instead of `Cancel` to ensure the continuations are not synchronously invoked.